config.w32 lacks config.m4's 64-bit guard; release.yml still generates x86 legs for the bundled build

## Summary

`config.w32` has no equivalent of `config.m4`'s 64-bit guard for the bundled
libJudy build, and `release.yml`'s Windows matrix is auto-generated with x86
legs. Post-vendoring (#142), the next release therefore runs ten x86 build
jobs against a bundled tree that cannot be built for 32-bit words.

## Background

The committed `libjudy/src/Judy1/Judy1Tables.c` and
`libjudy/src/JudyL/JudyLTables.c` are pre-generated under `-DJU_64BIT` and are
wrong for 32-bit words. `config.m4` refuses such a target at configure time:

> the bundled libJudy currently supports 64-bit targets only (its pre-generated
> tables assume JU_64BIT); use --with-judy=DIR to link a system libJudy instead

`config.w32` has no such check. It correctly omits `/D JU_64BIT` when `X64` is
false, and its own comment states the problem ("x86 builds must not get it, and
would need regenerated 32-bit tables besides"), but it then compiles the
64-bit-generated tables anyway.

## It is NOT a silent failure — verified

The initial suspicion was that an x86 bundled build would compile clean and
behave wrong (the #131 bug class). That is **wrong**, and worth recording so
nobody re-derives it:

`libjudy/src/wrappers/judy_static_asserts.c` does pass on x86 — its geometry
pins are under `#ifdef JU_64BIT` and the `#else` branch asserts
`sizeof(Word_t) == 4`, which x86 satisfies. But both **table files carry their
own pin**:

```c
// This file was generated under JU_64BIT: a 32-bit build must regenerate
// it, not reuse it.
JUDY_PHP_TABLES_ASSERT(sizeof(Word_t) == 8, word_t_is_8_bytes);
```

On MSVC x86 `_WIN64` is undefined, so `Judy.h` selects
`typedef unsigned long Word_t` (4 bytes) and that pin fails.

Reproduced locally by compiling the table files with `Word_t` forced to a
4-byte type and `JU_64BIT` undefined (clang 17, arm64) — 11 hard errors per
file, including:

```
Judy1Tables.c:166:1: error: 'judy_php_tables_assert_word_t_is_8_bytes'
        declared as an array with a negative size
Judy1Tables.c:170:1: error: 'judy_php_tables_assert_leaf2_maxpop1_is_128'
        declared as an array with a negative size
Judy1Tables.c:85:22: error: use of undeclared identifier 'cJ1_LEAF4_MAXPOP1'
```

So the defect is **diagnostics and release-matrix hygiene, not data corruption**.
No silently-broken x86 DLL can be produced; the build dies. But it dies with a
dozen cryptic C2118 "negative subscript" errors deep inside a generated
vendored source, rather than with the one-line configure message that Unix
gets.

## Exposure — verified, and it is real

`release.yml` generates its Windows matrix with
`php/php-windows-builder/extension-matrix@v1`, whose `ARCH_LIST` defaults to
`"x64,x86"`. The v2.5.2 release run (`32100574626`, 2026-08-18) produced
**10 x86 build jobs** alongside 11 x64 ones:

```
build-windows (windows-2022, 8.1, x86, nts, false) => success
build-windows (windows-2022, 8.2, x86, ts,  false) => success
build-windows (windows-2022, 8.3, x86, ts,  false) => success
...
```

Those succeeded because `v2.5.2`'s `config.w32` predates the vendoring (`git
show v2.5.2:config.w32` has no bundled branch) and linked a downloaded system
libJudy. On current `main` they would all fail.

Worse, `release.yml`'s `build-windows` job does not set `fail-fast: false`, so
the first x86 failure cancels the surviving x64 legs; and the `release` job
`needs: [build-windows, pecl-package]`, so it would be skipped entirely. **The
next release would publish no Windows DLLs and no PECL tarball at all.**

`ci.yml` is pinned `arch: ["x64"]`, so CI never builds x86 and would never
surface this before a release.

## Fix

1. **`config.w32`**: hard-fail the bundled configure via `ERROR()` when `!X64`,
   mirroring `config.m4`'s message. php-src sets `X64 = TARGET_ARCH != 'x86'`,
   so x64 and arm64 both pass and only 32-bit is rejected. `ERROR()` quits the
   configure; a `WARNING()` would not.
2. **`release.yml`**: pass `arch-list: x64` to `extension-matrix@v1` so the x86
   legs are never generated, keeping the release matrix in sync with `ci.yml`'s
   `arch: ["x64"]`.
3. **Defense in depth**: an `#ifndef JU_64BIT / #error` at the top of both table
   files, so the failure names its cause before the compiler reaches the
   64-bit-only size classes below — and so it holds for any build system, not
   just the two in this repo.

Note: the README/support-statement side is already covered by #158, which adds
"CI builds and tests the **x64** configuration" plus "The bundled build requires
a 64-bit target; on a 32-bit platform, use `--with-judy=DIR`".

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -906,6 +906,128 @@ jobs:
           Get-ChildItem -Path ".\build" -Recurse -Filter "build*.log" -ErrorAction SilentlyContinue |
             Select-Object -First 1 | ForEach-Object { Get-Content $_.FullName -Tail 30 }
 
+  # Negative control for the 32-bit bundled-build guards (#159). Inverted, in
+  # the same spirit as the differential-fuzz negative control (#142 Stage 4):
+  # this job is GREEN when the guards correctly REFUSE a 32-bit bundled build,
+  # and RED if a guard is ever removed, weakened, or its message drifts.
+  #
+  # Why this and not an x86 leg in build-windows: an x86 leg is red-when-correct,
+  # which reads as a broken board and invites someone to "fix" it by deleting the
+  # guard. The one-time proof that the config.w32 guard fires on real MSVC x86 is
+  # recorded in issue #159 (CI job 95934686125: configure aborts with
+  # "ERROR: the bundled libJudy currently supports 64-bit targets only ..."). What
+  # can silently drift afterwards is the guard's own source, and that is exactly
+  # what this job pins -- deterministically, on a Linux runner, in seconds.
+  #
+  # Each assertion matches the guard's MESSAGE, not merely a nonzero exit, so an
+  # unrelated breakage cannot masquerade as the guard working.
+  guard-32bit:
+    name: 32-bit bundled build is refused (negative control)
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v7
+
+      # config.w32 is JScript. Drive it through the php-src Windows build API,
+      # stubbed, with X64 set the way php-src sets it (win32/build/config.w32:
+      # "X64 = TARGET_ARCH != 'x86'"). ERROR() throws here, standing in for
+      # confutils.js's WScript.Quit(3).
+      - name: config.w32 refuses a 32-bit target
+        run: |
+          cat > /tmp/harness.js <<'HARNESS'
+          var fs = require('fs');
+          var st = { flags: [], errors: [], warnings: [], extensions: [] };
+          function ARG_WITH() {}
+          function EXTENSION() { st.extensions.push(Array.prototype.slice.call(arguments)); }
+          function ADD_EXTENSION_DEP() {}
+          function AC_DEFINE() {}
+          function ADD_FLAG(n, v) { st.flags.push(n + ' ' + v); }
+          function ADD_SOURCES() {}
+          function CHECK_LIB() { return false; }
+          function CHECK_HEADER_ADD_INCLUDE() { return false; }
+          function WARNING(m) { st.warnings.push(m); }
+          function ERROR(m) { st.errors.push(m); throw { __abort: true }; }
+          var configure_module_dirname = 'C:\\php\\pecl\\judy';
+          var PHP_JUDY = 'yes';
+          var PHP_JUDY_SHARED = true;
+          var X64 = (process.argv[3] !== 'x86');
+          try { eval(fs.readFileSync(process.argv[2], 'utf8')); }
+          catch (e) { if (!e || !e.__abort) throw e; }
+          console.log(JSON.stringify(st));
+          HARNESS
+
+          echo "--- x86: must ERROR with the 64-bit-only message ---"
+          node /tmp/harness.js config.w32 x86 > /tmp/x86.json
+          cat /tmp/x86.json
+          node -e '
+            var r = require("/tmp/x86.json");
+            var m = r.errors.join("\n");
+            if (r.errors.length !== 1) {
+              throw new Error("expected exactly 1 ERROR() on x86, got " + r.errors.length +
+                " -- the config.w32 32-bit guard is missing or no longer aborts");
+            }
+            if (!/bundled libJudy currently supports 64-bit targets only/.test(m) ||
+                !/JU_64BIT/.test(m) || !/--with-judy=DIR/.test(m)) {
+              throw new Error("ERROR() message drifted; it must still name the cause " +
+                "(64-bit only / JU_64BIT) and the way out (--with-judy=DIR). Got: " + m);
+            }
+            if (r.extensions.length !== 0) {
+              throw new Error("configure kept going after the guard fired");
+            }
+            console.log("OK: x86 refused with the expected message");
+          '
+
+          echo "--- x64: must proceed and define JU_64BIT ---"
+          node /tmp/harness.js config.w32 x64 > /tmp/x64.json
+          cat /tmp/x64.json
+          node -e '
+            var r = require("/tmp/x64.json");
+            if (r.errors.length !== 0) {
+              throw new Error("x64 must not be refused, got: " + r.errors.join("; "));
+            }
+            if (r.flags.indexOf("CFLAGS_JUDY /D JU_64BIT") === -1) {
+              throw new Error("x64 build lost /D JU_64BIT; flags were: " + r.flags.join(" | "));
+            }
+            if (r.extensions.length !== 1) {
+              throw new Error("x64 build did not register the extension");
+            }
+            console.log("OK: x64 proceeds with /D JU_64BIT");
+          '
+
+      # Backstop that lives with the data: the pre-generated tables must refuse
+      # to compile without JU_64BIT whatever build system wires them in.
+      - name: Pre-generated tables refuse a 32-bit word
+        run: |
+          # Simulate MSVC x86 / any ILP32 target: claim Judy.h's _WORD_T guard
+          # with a genuinely 4-byte Word_t and leave JU_64BIT undefined.
+          cat > /tmp/word32.h <<'EOF'
+          #define _WORD_T
+          typedef unsigned int Word_t, * PWord_t;
+          EOF
+
+          INC="-Ilibjudy/src -Ilibjudy/src/JudyCommon -Ilibjudy/src/Judy1 -Ilibjudy/src/JudyL"
+          rc=0
+          for f in libjudy/src/Judy1/Judy1Tables.c libjudy/src/JudyL/JudyLTables.c; do
+            echo "--- $f: 32-bit word must be rejected ---"
+            if cc -fsyntax-only -include /tmp/word32.h $INC "$f" 2>/tmp/err.txt; then
+              echo "FAIL: $f compiled with a 4-byte Word_t and no JU_64BIT"; rc=1; continue
+            fi
+            if ! grep -q "require a 64-bit build" /tmp/err.txt; then
+              echo "FAIL: $f was rejected, but not by the JU_64BIT guard."
+              echo "      The #error backstop is missing or its message drifted."
+              head -20 /tmp/err.txt; rc=1; continue
+            fi
+            echo "OK: rejected by the JU_64BIT #error guard"
+
+            echo "--- $f: 64-bit build must still be clean ---"
+            if ! cc -c -Wall -Werror -O2 -DJU_64BIT $INC "$f" -o /tmp/ok.o 2>/tmp/err64.txt; then
+              echo "FAIL: $f no longer compiles for a normal 64-bit build"
+              head -20 /tmp/err64.txt; rc=1; continue
+            fi
+            echo "OK: 64-bit build unaffected"
+          done
+          exit $rc
+
   report-results:
     needs: [build-linux, build-windows]
     if: ${{ !cancelled() }}

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -42,8 +42,17 @@ jobs:
       - name: Get the extension matrix
         id: extension-matrix
         uses: php/php-windows-builder/extension-matrix@v1
+        with:
+          # The action defaults to arch-list "x64,x86". The bundled libJudy
+          # is 64-bit-only -- its pre-generated Judy1Tables.c/JudyLTables.c
+          # assume JU_64BIT -- so config.w32 hard-fails a 32-bit configure.
+          # Left unconstrained, every x86 leg would fail and (with the
+          # default fail-fast) take the whole release down, publishing no
+          # DLLs at all. x64 is what this project supports and tests; keep
+          # this list in sync with ci.yml's build-windows matrix.
+          arch-list: x64
 
-  # Build Windows DLLs for every PHP version / arch / thread-safety combination
+  # Build Windows DLLs for every PHP version / thread-safety combination (x64)
   build-windows:
     needs: get-extension-matrix
     runs-on: ${{ matrix.os }}

--- a/config.w32
+++ b/config.w32
@@ -8,6 +8,21 @@ if (PHP_JUDY != "no") {
 		// Bundled libJudy (libjudy/, Judy-1.0.5 + the documented patch
 		// series -- see libjudy/PATCHES.md). The P5 LLP64 patches are
 		// real source diffs now, so no download/patch step is needed.
+
+		// The bundled build is 64-bit-only: the committed
+		// Judy1Tables.c/JudyLTables.c were pre-generated under JU_64BIT
+		// and are wrong for 32-bit words. Refuse the target here, the
+		// way config.m4 does on unix, instead of letting the build run
+		// on and die inside a generated table file on its
+		// sizeof(Word_t) == 8 pin -- correct, but a dozen cryptic C2118
+		// errors deep in vendored sources. php-src sets
+		// X64 = TARGET_ARCH != 'x86', so x64 and arm64 both pass here
+		// and only a 32-bit target is rejected. ERROR() aborts the
+		// configure; a WARNING would still emit a broken DLL.
+		if (!X64) {
+			ERROR("the bundled libJudy currently supports 64-bit targets only (its pre-generated tables assume JU_64BIT); use --with-judy=DIR to link a system libJudy instead");
+		}
+
 		EXTENSION("judy", "php_judy.c judy_arrayaccess.c judy_handlers.c judy_iterator.c");
 		ADD_EXTENSION_DEP("judy", "json");
 		AC_DEFINE('HAVE_JUDY', 1, 'Have Judy library');

--- a/libjudy/src/Judy1/Judy1Tables.c
+++ b/libjudy/src/Judy1/Judy1Tables.c
@@ -162,7 +162,14 @@ j__1_LeafWPopToWords[cJ1_LEAFW_MAXPOP1 + 1] =
         typedef char judy_php_tables_assert_ ## name [(cond) ? 1 : -1]
 
 // This file was generated under JU_64BIT: a 32-bit build must regenerate
-// it, not reuse it.
+// it, not reuse it.  The #error states the cause in one line; the pin below
+// is the belt to its suspenders, and still catches a build that forces
+// JU_64BIT on despite a 32-bit word.  config.m4 and config.w32 both refuse a
+// 32-bit target before reaching this file -- these two are the backstop for
+// any other build system, and for a hand-rolled compile of these sources.
+#ifndef JU_64BIT
+#error "php-judy: these pre-generated libJudy tables require a 64-bit build (-DJU_64BIT). Regenerate them for 32-bit words, or link a system libJudy instead (--with-judy=DIR)."
+#endif
 JUDY_PHP_TABLES_ASSERT(sizeof(Word_t) == 8, word_t_is_8_bytes);
 
 // Dimensions the generator baked into the initializer lists above.

--- a/libjudy/src/JudyL/JudyLTables.c
+++ b/libjudy/src/JudyL/JudyLTables.c
@@ -247,7 +247,14 @@ j__L_LeafVPopToWords[cJU_BITSPERSUBEXPL + 1] =
         typedef char judy_php_tables_assert_ ## name [(cond) ? 1 : -1]
 
 // This file was generated under JU_64BIT: a 32-bit build must regenerate
-// it, not reuse it.
+// it, not reuse it.  The #error states the cause in one line; the pin below
+// is the belt to its suspenders, and still catches a build that forces
+// JU_64BIT on despite a 32-bit word.  config.m4 and config.w32 both refuse a
+// 32-bit target before reaching this file -- these two are the backstop for
+// any other build system, and for a hand-rolled compile of these sources.
+#ifndef JU_64BIT
+#error "php-judy: these pre-generated libJudy tables require a 64-bit build (-DJU_64BIT). Regenerate them for 32-bit words, or link a system libJudy instead (--with-judy=DIR)."
+#endif
 JUDY_PHP_TABLES_ASSERT(sizeof(Word_t) == 8, word_t_is_8_bytes);
 
 // Dimensions the generator baked into the initializer lists above.

